src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs

```diff
@@ -427,17 +431,34 @@ HashSet<ITypeSymbol> typeStack
 			// We have to walk all base types and interfaces to find the
 			// type param in the same position and examine those to see if
 			// the immutability attribute is present.
-			var parameterIndex = typeParameter.ContainingType.TypeParameters.IndexOf( typeParameter );
 			var currentType = typeParameter.ContainingType;
 			while( currentType != null ) {
 				// Check all interfaces
 				foreach( var intf in currentType.Interfaces ) {
-					if( InspectInterfaceTypeParameter( parameterIndex, intf ) ) {
+					// Find the matching type argument in the base interface
+					// that corresponds to this type parameter.  That is,
+					// if we have Foo<S, T>: IFoo<S>, IBar<T>, this will
+					// match up the Foo S, to the IFoo S, but will get -1
+					// from IBar since it doesn't have S.
+					var ordinal = intf.TypeArguments
+						.IndexOf( intf.TypeArguments
+							.FirstOrDefault( tp => tp.Name == typeParameter.Name ) );
+
+					// If it doesn't have a matching type argument then the
+					// type we're inspecting isn't defined in this interface.
+					if ( ordinal < 0) {
+						continue;
+					}
+
+					// Grab the implemented interfaces version of the type 
+					// argument and then pass it down for inspection
+					var subTypeParameter = intf.TypeArguments[ordinal] as ITypeParameterSymbol;

```

Comment:
nit: It feels wrong to find the typeParameter (or default), then throw it away to get the ordinal, then use the ordinal to get the original typeArgument back (`intf.TypeArguments[ordinal]`).  It makes more intuitive sense to me to do something like:
```
var matchingArgument = intf.TypeArguments.FirstOrDefault( ... );
if ( matchingArgument == default(TypeArgument) ) { continue; }
var ordinal = intf.TypeArguments.IndexOf( matchingArgument );
```
but that is a really tiny nitpick and it probably won't make any difference in the long run